Cabal/src/Distribution/Simple/ShowBuildInfo.hs

```diff
@@ -70,85 +75,110 @@ import Distribution.Simple.LocalBuildInfo
-      ]
-      where
-        path = maybe JsonNull (JsonString . programPath)
-               $ (flavorToProgram . compilerFlavor $ compiler lbi)
-               >>= flip lookupProgram (withPrograms lbi)
+mkCompilerInfo :: ProgramDb -> Compiler -> Json
+mkCompilerInfo programDb cmplr = JsonObject
+  [ "flavour"     .= JsonString (T.pack (prettyShow $ compilerFlavor cmplr))
+  , "compiler-id" .= JsonString (T.pack (showCompilerId cmplr))
+  , "path"        .= path
+  ]
+  where
+    path = maybe JsonNull (JsonString . T.pack . programPath)
+            $ (flavorToProgram . compilerFlavor $ cmplr)
+            >>= flip lookupProgram programDb
 
-        flavorToProgram :: CompilerFlavor -> Maybe Program
-        flavorToProgram GHC   = Just ghcProgram
-        flavorToProgram GHCJS = Just ghcjsProgram
-        flavorToProgram UHC   = Just uhcProgram
-        flavorToProgram JHC   = Just jhcProgram
-        flavorToProgram _     = Nothing
+    flavorToProgram :: CompilerFlavor -> Maybe Program
+    flavorToProgram GHC   = Just ghcProgram
+    flavorToProgram GHCJS = Just ghcjsProgram
+    flavorToProgram UHC   = Just uhcProgram
+    flavorToProgram JHC   = Just jhcProgram
+    flavorToProgram _     = Nothing
 
-    mkComponentInfo (name, clbi) = JsonObject
-      [ "type"          .= JsonString compType
-      , "name"          .= JsonString (prettyShow name)
-      , "unit-id"       .= JsonString (prettyShow $ componentUnitId clbi)
-      , "compiler-args" .= JsonArray (map JsonString $ getCompilerArgs bi lbi clbi)
-      , "modules"       .= JsonArray (map (JsonString . display) modules)
-      , "src-files"     .= JsonArray (map JsonString sourceFiles)
-      , "src-dirs"      .= JsonArray (map JsonString $ map getSymbolicPath $ hsSourceDirs bi)
-      ]
-      where
-        bi = componentBuildInfo comp
-        comp = fromMaybe (error $ "mkBuildInfo: no component " ++ prettyShow name) $ lookupComponent pkg_descr name
-        compType = case comp of
-          CLib _   -> "lib"
-          CExe _   -> "exe"
-          CTest _  -> "test"
-          CBench _ -> "bench"
-          CFLib _  -> "flib"
-        modules = case comp of
-          CLib lib -> explicitLibModules lib
-          CExe exe -> exeModules exe
-          _        -> []
-        sourceFiles = case comp of
-          CLib _   -> []
-          CExe exe -> [modulePath exe]
-          _        -> []
+mkComponentInfo :: FilePath -> PackageDescription -> LocalBuildInfo -> ComponentLocalBuildInfo -> Json
+mkComponentInfo wdir pkg_descr lbi clbi = JsonObject $
+  [ "type"          .= JsonString compType
+  , "name"          .= JsonString (T.pack $ prettyShow name)
+  , "unit-id"       .= JsonString (T.pack $ prettyShow $ componentUnitId clbi)
+  , "compiler-args" .= JsonArray (map JsonString $ getCompilerArgs bi lbi clbi)
+  , "modules"       .= JsonArray (map (JsonString . T.pack . display) modules)
+  , "src-files"     .= JsonArray (map (JsonString . T.pack) sourceFiles)
```

Comment:
this is not ideal as `sourceFiles` are relative to `hs-src-dirs`. Getting the final Filepath for the source file requires more post-processing in tooling.